File: hariom-erp/services/spec-service/src/spec_math.py
"""Canonical spec-sheet math.

Authoritative implementation. Mirror in TypeScript at
`apps/web-ui/lib/spec-math.ts`. Any change here must be ported there and
covered by tests on both sides.

All formulas are documented in `IMPLEMENTATION.md` (repo root).
"""
from __future__ import annotations

import math
from dataclasses import dataclass, field
from typing import Iterable, List, Optional, Sequence

# ---------------------------------------------------------------------------
# Constants (defaults; per-spec overrides allowed in the Validation footer)
# ---------------------------------------------------------------------------

GLOBAL_ADHESIVE_PERCENT: float = 15.0
GLOBAL_PARCHMENT_PERCENT: float = 1.5
GLOBAL_MOISTURE_LOSS_PERCENT: float = 9.0

BAMBOO_LENGTH_MIN_MM: int = 1390
BAMBOO_LENGTH_MAX_MM: int = 1560
BAMBOO_LENGTH_STEP_MM: int = 10
BAMBOO_CUT_LOSS_MM: int = 40

MANDREL_TOLERANCE_MM: float = 0.1

RECIPE_MIN_PAPERS: int = 3
RECIPE_MAX_PAPERS: int = 5
RECIPE_MAX_PLIES: int = 18

DELTA_ABS_G: float = 3.0
DELTA_PCT: float = 0.0


# ---------------------------------------------------------------------------
# Types
# ---------------------------------------------------------------------------

@dataclass(frozen=True)
class RecipePaper:
    """One paper in the recipe, repeated `ply_count` times in order."""

    paper_id: str
    gsm: float
    bulk: float
    ply_count: int
    code: str = ""


@dataclass(frozen=True)
class WeightBreakdown:
    paper_g: float
    adhesive_g: float
    parchment_g: float
    wet_g: float
    dry_g: float


@dataclass(frozen=True)
class BambooPlan:
    bamboo_length_mm: int
    tubes_per_bamboo: int
    trim_waste_mm: int
    usable_length_mm: int
    finished_length_mm: int
    fixed_end_trim_mm: int
    residual_offcut_mm: int
    total_trim_mm: int


@dataclass(frozen=True)
class RecipeValidation:
    distinct_papers: int
    total_plies: int
    papers_ok: bool
    plies_ok: bool
    delta_g: float
    delta_tolerance_g: float
    delta_ok: bool

    @property
    def ok(self) -> bool:
        return self.papers_ok and self.plies_ok and self.delta_ok


@dataclass(frozen=True)
class PreviewResult:
    id_mm: float
    od_mm: float
    wall_mm: float
    per_ply_thickness_mm: List[float]
    per_ply_avg_dia_mm: List[float]
    per_ply_weight_per_mm_g: List[float]
    target_per_ply_weight_per_mm_g: List[float]
    paper_weight_per_mm_g: float
    target_paper_weight_per_mm_g: float
    tube: WeightBreakdown
    nominal_tube: WeightBreakdown
    bamboo: WeightBreakdown
    bamboo_trim: WeightBreakdown
    whole_bamboo: WeightBreakdown
    bamboo_plan: BambooPlan
    paper_required_g: float
    paper_calibration_factor: float
    validation: RecipeValidation


# ---------------------------------------------------------------------------
# Core formulas
# ---------------------------------------------------------------------------

def thickness_mm(gsm: float, bulk: float) -> float:
    """`thickness_mm = GSM × bulk / 1000`.

    GSM in g/m², bulk in cm³/g. Handles zero / negative inputs by clamping.
    """
    if gsm is None or bulk is None:
        return 0.0
    return max(float(gsm), 0.0) * max(float(bulk), 0.0) / 1000.0


def expand_plies(papers: Sequence[RecipePaper]) -> List[RecipePaper]:
    """Expand `ply_count` so `papers[i]` is the i-th ply from inner to outer.

    Each returned entry has `ply_count == 1` and the same paper identity.
    """
    out: List[RecipePaper] = []
    for p in papers:
        for _ in range(max(int(p.ply_count or 0), 0)):
            out.append(
                RecipePaper(
                    paper_id=p.paper_id,
                    gsm=p.gsm,
                    bulk=p.bulk,
                    ply_count=1,
                    code=p.code,
                )
            )
    return out


def ply_geometry(
    id_mm: float, expanded: Sequence[RecipePaper]
) -> tuple[List[float], List[float], float]:
    """Return (per-ply thickness, per-ply avg_dia, wall_mm).

    Plies are iterated from inner (ply 1) to outer (ply N); wall is the sum
    of all thicknesses.
    """
    thicknesses: List[float] = []
    avg_dias: List[float] = []
    cumulative_inner: float = 0.0
    for p in expanded:
        t = thickness_mm(p.gsm, p.bulk)
        inner_d = id_mm + 2.0 * cumulative_inner
        avg_d = inner_d + t  # (inner + (inner + 2t)) / 2 = inner + t
        thicknesses.append(t)
        avg_dias.append(avg_d)
        cumulative_inner += t
    return thicknesses, avg_dias, cumulative_inner


def per_ply_weight_per_mm(gsm: float, avg_dia_mm: float) -> float:
    """Grams per mm of tube length for one ply.

    Surface area (m²) for 1 mm of tube = π × d(mm) × 1 / 1_000_000. Mass = GSM × area.
    """
    return max(float(gsm), 0.0) * math.pi * max(float(avg_dia_mm), 0.0) / 1_000_000.0


def dry_divisor(moisture_loss_percent: float = GLOBAL_MOISTURE_LOSS_PERCENT) -> float:
    return max(1.0 - max(float(moisture_loss_percent or 0.0), 0.0) / 100.0, 0.0001)


def addon_share(
    adhesive_percent: float = GLOBAL_ADHESIVE_PERCENT,
    parchment_percent: float = GLOBAL_PARCHMENT_PERCENT,
    parchment_allowed: bool = True,
) -> float:
    adhesive_share = max(float(adhesive_percent), 0.0) / 100.0
    parchment_share = max(float(parchment_percent), 0.0) / 100.0 if parchment_allowed else 0.0
    return adhesive_share + parchment_share


def wet_dry_breakdown(
    paper_g: float,
    *,
    adhesive_percent: float = GLOBAL_ADHESIVE_PERCENT,
    parchment_percent: float = GLOBAL_PARCHMENT_PERCENT,
    moisture_loss_percent: float = GLOBAL_MOISTURE_LOSS_PERCENT,
    parchment_allowed: bool = True,
    target_dry_g: float = 0.0,
) -> WeightBreakdown:
    paper_g = max(float(paper_g), 0.0)
    target_dry = max(float(target_dry_g or 0.0), 0.0)
    divisor = dry_divisor(moisture_loss_percent)
    addons = addon_share(adhesive_percent, parchment_percent, parchment_allowed)
    dry_base_g = target_dry if target_dry > 0 else paper_g * divisor / max(1.0 - divisor * addons, 0.0001)
    adhesive_g = dry_base_g * max(float(adhesive_percent), 0.0) / 100.0
    parchment_g = dry_base_g * max(float(parchment_percent), 0.0) / 100.0 if parchment_allowed else 0.0
    wet_g = paper_g + adhesive_g + parchment_g
    dry_g = wet_g * divisor
    return WeightBreakdown(
        paper_g=round(paper_g, 4),
        adhesive_g=round(adhesive_g, 4),
        parchment_g=round(parchment_g, 4),
        wet_g=round(wet_g, 4),
        dry_g=round(dry_g, 4),
    )


def scale_breakdown(breakdown: WeightBreakdown, factor: float) -> WeightBreakdown:
    """Scale one finished-tube breakdown to another length/count basis."""
    scale = max(float(factor or 0.0), 0.0)
    return WeightBreakdown(
        paper_g=round(breakdown.paper_g * scale, 4),
        adhesive_g=round(breakdown.adhesive_g * scale, 4),
        parchment_g=round(breakdown.parchment_g * scale, 4),
        wet_g=round(breakdown.wet_g * scale, 4),
        dry_g=round(breakdown.dry_g * scale, 4),
    )


def required_paper_g(
    target_dry_g: float,
    *,
    adhesive_percent: float = GLOBAL_ADHESIVE_PERCENT,
    parchment_percent: float = GLOBAL_PARCHMENT_PERCENT,
    moisture_loss_percent: float = GLOBAL_MOISTURE_LOSS_PERCENT,
    parchment_allowed: bool = True,
) -> float:
    """Reverse: paper needed when adhesive/parchment are percentages of client dry weight."""
    if target_dry_g is None or target_dry_g <= 0:
        return 0.0
    target_dry = float(target_dry_g)
    wet_target_g = target_dry / dry_divisor(moisture_loss_percent)
    adhesive_g = target_dry * max(float(adhesive_percent), 0.0) / 100.0
    parchment_g = target_dry * max(float(parchment_percent), 0.0) / 100.0 if parchment_allowed else 0.0
    return round(max(wet_target_g - adhesive_g - parchment_g, 0.0), 4)


# ---------------------------------------------------------------------------
# Bamboo plan
# ---------------------------------------------------------------------------

def build_bamboo_plan(
    tube_length_mm: float,
    *,
    min_length_mm: int = BAMBOO_LENGTH_MIN_MM,
    max_length_mm: int = BAMBOO_LENGTH_MAX_MM,
    step_mm: int = BAMBOO_LENGTH_STEP_MM,
    cut_loss_mm: int = BAMBOO_CUT_LOSS_MM,
) -> BambooPlan:
    """Pick the bamboo length that maximises tubes first, then minimises waste."""
    tube_len = max(int(round(tube_length_mm)), 1)
    best: Optional[BambooPlan] = None
    for L in range(max_length_mm, min_length_mm - 1, -step_mm):
        usable = max(L - cut_loss_mm, 0)
        tubes = usable // tube_len if tube_len else 0
        waste = usable - tubes * tube_len
        candidate = BambooPlan(
            bamboo_length_mm=L,
            tubes_per_bamboo=tubes,
            trim_waste_mm=waste,
            usable_length_mm=usable,
            finished_length_mm=tubes * tube_len,
            fixed_end_trim_mm=min(max(cut_loss_mm, 0), L),
            residual_offcut_mm=waste,
            total_trim_mm=max(L - tubes * tube_len, 0),
        )
        if best is None:
            best = candidate
            continue
        # tubes desc, waste asc, length desc
        if candidate.tubes_per_bamboo > best.tubes_per_bamboo:
            best = candidate
        elif candidate.tubes_per_bamboo == best.tubes_per_bamboo and candidate.trim_waste_mm < best.trim_waste_mm:
            best = candidate
    assert best is not None
    return best


# ---------------------------------------------------------------------------
# Validation
# ---------------------------------------------------------------------------

def validate_recipe(
    papers: Sequence[RecipePaper],
    target_dry_g: float,
    predicted_dry_g: float,
) -> RecipeValidation:
    distinct = {p.paper_id for p in papers if int(p.ply_count or 0) > 0}
    total_plies = sum(max(int(p.ply_count or 0), 0) for p in papers)
    papers_ok = RECIPE_MIN_PAPERS <= len(distinct) <= RECIPE_MAX_PAPERS
    plies_ok = 1 <= total_plies <= RECIPE_MAX_PLIES
    delta = (predicted_dry_g or 0.0) - (target_dry_g or 0.0)
    tol = DELTA_ABS_G
    delta_ok = abs(delta) <= tol
    return RecipeValidation(
        distinct_papers=len(distinct),
        total_plies=total_plies,
        papers_ok=papers_ok,
        plies_ok=plies_ok,
        delta_g=round(delta, 4),
        delta_tolerance_g=round(tol, 4),
        delta_ok=delta_ok,
    )


# ---------------------------------------------------------------------------
# Top-level orchestration (what POST /specs/preview returns)
# ---------------------------------------------------------------------------

def compute_preview(
    *,
    mandrel_od_mm: float,
    tube_length_mm: float,
    papers: Sequence[RecipePaper],
    target_dry_g: float,
    adhesive_percent: float = GLOBAL_ADHESIVE_PERCENT,
    parchment_percent: float = GLOBAL_PARCHMENT_PERCENT,
    moisture_loss_percent: float = GLOBAL_MOISTURE_LOSS_PERCENT,
    parchment_allowed: bool = True,
) -> PreviewResult:
    id_mm = max(float(mandrel_od_mm or 0.0), 0.0)
    tube_len = max(float(tube_length_mm or 0.0), 0.0)

    expanded = expand_plies(papers)
    thicknesses, avg_dias, wall = ply_geometry(id_mm, expanded)
    per_ply_wpm = [
        per_ply_weight_per_mm(p.gsm, d) for p, d in zip(expanded, avg_dias)
    ]
    paper_wpm = sum(per_ply_wpm)
    od_mm = id_mm + 2.0 * wall

    nominal_tube_paper_g = paper_wpm * tube_len
    nominal_tube = wet_dry_breakdown(
        nominal_tube_paper_g,
        adhesive_percent=adhesive_percent,
        parchment_percent=parchment_percent,
        moisture_loss_percent=moisture_loss_percent,
        parchment_allowed=parchment_allowed,
        target_dry_g=target_dry_g,
    )

    required = required_paper_g(
        target_dry_g,
        adhesive_percent=adhesive_percent,
        parchment_percent=parchment_percent,
        moisture_loss_percent=moisture_loss_percent,
        parchment_allowed=parchment_allowed,
    )
    has_recipe = nominal_tube_paper_g > 0 and len(expanded) > 0
    allocated_tube_paper_g = required if has_recipe and float(target_dry_g or 0.0) > 0 else nominal_tube_paper_g
    paper_calibration_factor = (
        allocated_tube_paper_g / nominal_tube_paper_g if nominal_tube_paper_g > 0 else 0.0
    )
    target_per_ply_wpm = [w * paper_calibration_factor for w in per_ply_wpm]
    target_paper_wpm = sum(target_per_ply_wpm)
    tube = (
        wet_dry_breakdown(
            allocated_tube_paper_g,
            adhesive_percent=adhesive_percent,
            parchment_percent=parchment_percent,
            moisture_loss_percent=moisture_loss_percent,
            parchment_allowed=parchment_allowed,
            target_dry_g=target_dry_g,
        )
        if has_recipe
        else WeightBreakdown(0.0, 0.0, 0.0, 0.0, 0.0)
    )

    plan = build_bamboo_plan(tube_len if tube_len else 1.0)
    tube_basis_mm = max(tube_len, 1.0)
    bamboo = scale_breakdown(tube, plan.tubes_per_bamboo)
    bamboo_trim = scale_breakdown(tube, plan.total_trim_mm / tube_basis_mm)
    whole_bamboo = scale_breakdown(tube, plan.bamboo_length_mm / tube_basis_mm)

    validation = validate_recipe(papers, target_dry_g or 0.0, tube.dry_g)

    return PreviewResult(
        id_mm=round(id_mm, 4),
        od_mm=round(od_mm, 4),
        wall_mm=round(wall, 4),
        per_ply_thickness_mm=[round(t, 4) for t in thicknesses],
        per_ply_avg_dia_mm=[round(d, 4) for d in avg_dias],
        per_ply_weight_per_mm_g=[round(w, 6) for w in per_ply_wpm],
        target_per_ply_weight_per_mm_g=[round(w, 6) for w in target_per_ply_wpm],
        paper_weight_per_mm_g=round(paper_wpm, 6),
        target_paper_weight_per_mm_g=round(target_paper_wpm, 6),
        tube=tube,
        nominal_tube=nominal_tube,
        bamboo=bamboo,
        bamboo_trim=bamboo_trim,
        whole_bamboo=whole_bamboo,
        bamboo_plan=plan,
        paper_required_g=required,
        paper_calibration_factor=round(paper_calibration_factor, 6),
        validation=validation,
    )


# ---------------------------------------------------------------------------
# Serialisation helper for the HTTP layer
# ---------------------------------------------------------------------------

def preview_to_dict(p: PreviewResult) -> dict:
    return {
        "id_mm": p.id_mm,
        "mandrel_tolerance_mm": MANDREL_TOLERANCE_MM,
        "od_mm": p.od_mm,
        "wall_mm": p.wall_mm,
        "per_ply_thickness_mm": p.per_ply_thickness_mm,
        "per_ply_avg_dia_mm": p.per_ply_avg_dia_mm,
        "per_ply_weight_per_mm_g": p.per_ply_weight_per_mm_g,
        "target_per_ply_weight_per_mm_g": p.target_per_ply_weight_per_mm_g,
        "paper_weight_per_mm_g": p.paper_weight_per_mm_g,
        "target_paper_weight_per_mm_g": p.target_paper_weight_per_mm_g,
        "tube": {
            "paper_g": p.tube.paper_g,
            "adhesive_g": p.tube.adhesive_g,
            "parchment_g": p.tube.parchment_g,
            "wet_g": p.tube.wet_g,
            "dry_g": p.tube.dry_g,
        },
        "nominal_tube": {
            "paper_g": p.nominal_tube.paper_g,
            "adhesive_g": p.nominal_tube.adhesive_g,
            "parchment_g": p.nominal_tube.parchment_g,
            "wet_g": p.nominal_tube.wet_g,
            "dry_g": p.nominal_tube.dry_g,
        },
        "bamboo": {
            "paper_g": p.bamboo.paper_g,
            "adhesive_g": p.bamboo.adhesive_g,
            "parchment_g": p.bamboo.parchment_g,
            "wet_g": p.bamboo.wet_g,
            "dry_g": p.bamboo.dry_g,
        },
        "bamboo_trim": {
            "paper_g": p.bamboo_trim.paper_g,
            "adhesive_g": p.bamboo_trim.adhesive_g,
            "parchment_g": p.bamboo_trim.parchment_g,
            "wet_g": p.bamboo_trim.wet_g,
            "dry_g": p.bamboo_trim.dry_g,
        },
        "whole_bamboo": {
            "paper_g": p.whole_bamboo.paper_g,
            "adhesive_g": p.whole_bamboo.adhesive_g,
            "parchment_g": p.whole_bamboo.parchment_g,
            "wet_g": p.whole_bamboo.wet_g,
            "dry_g": p.whole_bamboo.dry_g,
        },
        "bamboo_plan": {
            "bamboo_length_mm": p.bamboo_plan.bamboo_length_mm,
            "tubes_per_bamboo": p.bamboo_plan.tubes_per_bamboo,
            "trim_waste_mm": p.bamboo_plan.trim_waste_mm,
            "usable_length_mm": p.bamboo_plan.usable_length_mm,
            "finished_length_mm": p.bamboo_plan.finished_length_mm,
            "fixed_end_trim_mm": p.bamboo_plan.fixed_end_trim_mm,
            "residual_offcut_mm": p.bamboo_plan.residual_offcut_mm,
            "total_trim_mm": p.bamboo_plan.total_trim_mm,
        },
        "paper_required_g": p.paper_required_g,
        "paper_calibration_factor": p.paper_calibration_factor,
        "validation": {
            "distinct_papers": p.validation.distinct_papers,
            "total_plies": p.validation.total_plies,
            "papers_ok": p.validation.papers_ok,
            "plies_ok": p.validation.plies_ok,
            "delta_g": p.validation.delta_g,
            "delta_tolerance_g": p.validation.delta_tolerance_g,
            "delta_ok": p.validation.delta_ok,
            "ok": p.validation.ok,
        },
        "constants": {
            "adhesive_percent": adhesive_percent_value(),
            "parchment_percent": parchment_percent_value(),
            "moisture_loss_percent": moisture_loss_percent_value(),
            "bamboo_length_min_mm": BAMBOO_LENGTH_MIN_MM,
            "bamboo_length_max_mm": BAMBOO_LENGTH_MAX_MM,
            "bamboo_length_step_mm": BAMBOO_LENGTH_STEP_MM,
            "bamboo_cut_loss_mm": BAMBOO_CUT_LOSS_MM,
            "mandrel_tolerance_mm": MANDREL_TOLERANCE_MM,
            "recipe_min_papers": RECIPE_MIN_PAPERS,
            "recipe_max_papers": RECIPE_MAX_PAPERS,
            "recipe_max_plies": RECIPE_MAX_PLIES,
        },
    }


# Small accessor shims so the HTTP layer can surface the "defaults" while the
# preview call used possibly-overridden values. These are just the module
# constants; kept as functions in case future runtime config lookup is added.

def adhesive_percent_value() -> float:
    return GLOBAL_ADHESIVE_PERCENT


def parchment_percent_value() -> float:
    return GLOBAL_PARCHMENT_PERCENT


def moisture_loss_percent_value() -> float:
    return GLOBAL_MOISTURE_LOSS_PERCENT

File: apps/web-ui/lib/spec-math.ts
/**
 * Canonical spec-sheet math — TypeScript mirror.
 *
 * Mirror of `hariom-erp/services/spec-service/src/spec_math.py`. Any change
 * here must be ported there and covered on both sides. See
 * `IMPLEMENTATION.md` in the repo root for the formula reference.
 */

// ---------------------------------------------------------------------------
// Constants (defaults; per-spec overrides allowed in the Validation footer)
// ---------------------------------------------------------------------------

export const GLOBAL_ADHESIVE_PERCENT = 15.0
export const GLOBAL_PARCHMENT_PERCENT = 1.5
export const GLOBAL_MOISTURE_LOSS_PERCENT = 9.0

export const BAMBOO_LENGTH_MIN_MM = 1390
export const BAMBOO_LENGTH_MAX_MM = 1560
export const BAMBOO_LENGTH_STEP_MM = 10
export const BAMBOO_CUT_LOSS_MM = 40

export const MANDREL_TOLERANCE_MM = 0.1

export const RECIPE_MIN_PAPERS = 1
export const RECIPE_MAX_PAPERS = 10
export const RECIPE_MAX_PLIES = 25

export const DELTA_ABS_G = 3.0
export const DELTA_PCT = 0.0

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export type RecipePaper = {
  paper_id: string
  gsm: number
  bulk: number
  ply_count: number
  code?: string
}

export type WeightBreakdown = {
  paper_g: number
  adhesive_g: number
  parchment_g: number
  wet_g: number
  dry_g: number
}

export type BambooPlan = {
  bamboo_length_mm: number
  tubes_per_bamboo: number
  trim_waste_mm: number
  usable_length_mm: number
  finished_length_mm: number
  fixed_end_trim_mm: number
  residual_offcut_mm: number
  total_trim_mm: number
}

export type RecipeValidation = {
  distinct_papers: number
  total_plies: number
  papers_ok: boolean
  plies_ok: boolean
  delta_g: number
  delta_tolerance_g: number
  delta_ok: boolean
  ok: boolean
}

export type PreviewResult = {
  id_mm: number
  mandrel_tolerance_mm: number
  od_mm: number
  wall_mm: number
  per_ply_thickness_mm: number[]
  per_ply_avg_dia_mm: number[]
  per_ply_weight_per_mm_g: number[]
  target_per_ply_weight_per_mm_g: number[]
  paper_weight_per_mm_g: number
  target_paper_weight_per_mm_g: number
  tube: WeightBreakdown
  nominal_tube: WeightBreakdown
  bamboo: WeightBreakdown
  bamboo_trim: WeightBreakdown
  whole_bamboo: WeightBreakdown
  bamboo_plan: BambooPlan
  paper_required_g: number
  paper_calibration_factor: number
  validation: RecipeValidation
}

export type PreviewOptions = {
  mandrel_od_mm: number
  tube_length_mm: number
  papers: RecipePaper[]
  target_dry_g: number
  adhesive_percent?: number
  parchment_percent?: number
  moisture_loss_percent?: number
  parchment_allowed?: boolean
}

// ---------------------------------------------------------------------------
// Core formulas
// ---------------------------------------------------------------------------

const round4 = (n: number) => Math.round(n * 10_000) / 10_000
const round6 = (n: number) => Math.round(n * 1_000_000) / 1_000_000
const clamp0 = (n: number | null | undefined) =>
  typeof n === 'number' && Number.isFinite(n) && n > 0 ? n : 0
const dryDivisor = (moistureLossPercent: number) => Math.max(1 - Math.max(moistureLossPercent, 0) / 100, 0.0001)
const addOnShare = (adhesivePercent: number, parchmentPercent: number, parchmentAllowed: boolean) =>
  Math.max(adhesivePercent, 0) / 100 + (parchmentAllowed ? Math.max(parchmentPercent, 0) / 100 : 0)

export function thicknessMm(gsm: number, bulk: number): number {
  return (clamp0(gsm) * clamp0(bulk)) / 1000
}

export function expandPlies(papers: readonly RecipePaper[]): RecipePaper[] {
  const out: RecipePaper[] = []
  for (const p of papers) {
    const count = Math.max(Math.floor(p.ply_count || 0), 0)
    for (let i = 0; i < count; i++) {
      out.push({
        paper_id: p.paper_id,
        gsm: p.gsm,
        bulk: p.bulk,
        ply_count: 1,
        code: p.code,
      })
    }
  }
  return out
}

export function plyGeometry(
  idMm: number,
  expanded: readonly RecipePaper[],
): { thicknesses: number[]; avg_dias: number[]; wall_mm: number } {
  const thicknesses: number[] = []
  const avg_dias: number[] = []
  let cumulativeInner = 0
  for (const p of expanded) {
    const t = thicknessMm(p.gsm, p.bulk)
    const innerD = idMm + 2 * cumulativeInner
    const avgD = innerD + t // (inner + (inner + 2t)) / 2 = inner + t
    thicknesses.push(t)
    avg_dias.push(avgD)
    cumulativeInner += t
  }
  return { thicknesses, avg_dias, wall_mm: cumulativeInner }
}

export function perPlyWeightPerMm(gsm: number, avgDiaMm: number): number {
  return (clamp0(gsm) * Math.PI * clamp0(avgDiaMm)) / 1_000_000
}

export function wetDryBreakdown(
  paperG: number,
  opts?: {
    adhesive_percent?: number
    parchment_percent?: number
    moisture_loss_percent?: number
    parchment_allowed?: boolean
    target_dry_g?: number
  },
): WeightBreakdown {
  const paper = clamp0(paperG)
  const A = opts?.adhesive_percent ?? GLOBAL_ADHESIVE_PERCENT
  const P = opts?.parchment_percent ?? GLOBAL_PARCHMENT_PERCENT
  const M = opts?.moisture_loss_percent ?? GLOBAL_MOISTURE_LOSS_PERCENT
  const parchmentAllowed = opts?.parchment_allowed ?? true
  const targetDry = clamp0(opts?.target_dry_g)
  const divisor = dryDivisor(M)
  const addon = addOnShare(A, P, parchmentAllowed)
  const dryBase = targetDry > 0 ? targetDry : (paper * divisor) / Math.max(1 - divisor * addon, 0.0001)
  const adhesive = (dryBase * Math.max(A, 0)) / 100
  const parchment = parchmentAllowed ? (dryBase * Math.max(P, 0)) / 100 : 0
  const wet = paper + adhesive + parchment
  const dry = wet * divisor
  return {
    paper_g: round4(paper),
    adhesive_g: round4(adhesive),
    parchment_g: round4(parchment),
    wet_g: round4(wet),
    dry_g: round4(dry),
  }
}

export function scaleBreakdown(breakdown: WeightBreakdown, factor: number): WeightBreakdown {
  const scale = Math.max(Number.isFinite(factor) ? factor : 0, 0)
  return {
    paper_g: round4(breakdown.paper_g * scale),
    adhesive_g: round4(breakdown.adhesive_g * scale),
    parchment_g: round4(breakdown.parchment_g * scale),
    wet_g: round4(breakdown.wet_g * scale),
    dry_g: round4(breakdown.dry_g * scale),
  }
}

export function requiredPaperG(
  targetDryG: number,
  opts?: {
    adhesive_percent?: number
    parchment_percent?: number
    moisture_loss_percent?: number
    parchment_allowed?: boolean
  },
): number {
  if (!targetDryG || targetDryG <= 0) return 0
  const A = opts?.adhesive_percent ?? GLOBAL_ADHESIVE_PERCENT
  const P = opts?.parchment_percent ?? GLOBAL_PARCHMENT_PERCENT
  const M = opts?.moisture_loss_percent ?? GLOBAL_MOISTURE_LOSS_PERCENT
  const parchmentAllowed = opts?.parchment_allowed ?? true
  const divisor = dryDivisor(M)
  const targetWet = targetDryG / divisor
  const adhesive = (targetDryG * Math.max(A, 0)) / 100
  const parchment = parchmentAllowed ? (targetDryG * Math.max(P, 0)) / 100 : 0
  return round4(Math.max(targetWet - adhesive - parchment, 0))
}

// ---------------------------------------------------------------------------
// Bamboo plan
// ---------------------------------------------------------------------------

export function buildBambooPlan(
  tubeLengthMm: number,
  opts?: {
    min_length_mm?: number
    max_length_mm?: number
    step_mm?: number
    cut_loss_mm?: number
  },
): BambooPlan {
  const tubeLen = Math.max(Math.round(tubeLengthMm), 1)
  const min = opts?.min_length_mm ?? BAMBOO_LENGTH_MIN_MM
  const max = opts?.max_length_mm ?? BAMBOO_LENGTH_MAX_MM
  const step = opts?.step_mm ?? BAMBOO_LENGTH_STEP_MM
  const cut = opts?.cut_loss_mm ?? BAMBOO_CUT_LOSS_MM
  let best: BambooPlan | null = null
  for (let L = max; L >= min; L -= step) {
    const usable = Math.max(L - cut, 0)
    const tubes = tubeLen > 0 ? Math.floor(usable / tubeLen) : 0
    const waste = usable - tubes * tubeLen
    const candidate: BambooPlan = {
      bamboo_length_mm: L,
      tubes_per_bamboo: tubes,
      trim_waste_mm: waste,
      usable_length_mm: usable,
      finished_length_mm: tubes * tubeLen,
      fixed_end_trim_mm: Math.min(Math.max(cut, 0), L),
      residual_offcut_mm: waste,
      total_trim_mm: Math.max(L - tubes * tubeLen, 0),
    }
    if (!best) {
      best = candidate
      continue
    }
    if (candidate.tubes_per_bamboo > best.tubes_per_bamboo) best = candidate
    else if (
      candidate.tubes_per_bamboo === best.tubes_per_bamboo &&
      candidate.trim_waste_mm < best.trim_waste_mm
    )
      best = candidate
  }
  return best!
}

// ---------------------------------------------------------------------------
// Validation
// ---------------------------------------------------------------------------

export function validateRecipe(
  papers: readonly RecipePaper[],
  targetDryG: number,
  predictedDryG: number,
): RecipeValidation {
  const distinct = new Set<string>()
  let totalPlies = 0
  for (const p of papers) {
    const count = Math.max(Math.floor(p.ply_count || 0), 0)
    if (count > 0) distinct.add(p.paper_id)
    totalPlies += count
  }
  const papers_ok = distinct.size >= RECIPE_MIN_PAPERS && distinct.size <= RECIPE_MAX_PAPERS
  const plies_ok = totalPlies >= 1 && totalPlies <= RECIPE_MAX_PLIES
  const delta = (predictedDryG || 0) - (targetDryG || 0)
  const tol = DELTA_ABS_G
  const delta_ok = Math.abs(delta) <= tol
  return {
    distinct_papers: distinct.size,
    total_plies: totalPlies,
    papers_ok,
    plies_ok,
    delta_g: round4(delta),
    delta_tolerance_g: round4(tol),
    delta_ok,
    ok: papers_ok && plies_ok && delta_ok,
  }
}

// ---------------------------------------------------------------------------
// Top-level orchestration
// ---------------------------------------------------------------------------

export function computePreview(opts: PreviewOptions): PreviewResult {
  const id_mm = clamp0(opts.mandrel_od_mm)
  const tube_length_mm = clamp0(opts.tube_length_mm)
  const adhesive_percent = opts.adhesive_percent ?? GLOBAL_ADHESIVE_PERCENT
  const parchment_percent = opts.parchment_percent ?? GLOBAL_PARCHMENT_PERCENT
  const moisture_loss_percent = opts.moisture_loss_percent ?? GLOBAL_MOISTURE_LOSS_PERCENT
  const parchment_allowed = opts.parchment_allowed ?? true

  const expanded = expandPlies(opts.papers)
  const { thicknesses, avg_dias, wall_mm } = plyGeometry(id_mm, expanded)
  const per_ply_wpm = expanded.map((p, i) => perPlyWeightPerMm(p.gsm, avg_dias[i]))
  const paper_wpm = per_ply_wpm.reduce((a, b) => a + b, 0)
  const od_mm = id_mm + 2 * wall_mm

  const nominal_tube_paper_g = paper_wpm * tube_length_mm
  const nominal_tube = wetDryBreakdown(nominal_tube_paper_g, {
    adhesive_percent,
    parchment_percent,
    moisture_loss_percent,
    parchment_allowed,
    target_dry_g: opts.target_dry_g,
  })

  const required = requiredPaperG(opts.target_dry_g, {
    adhesive_percent,
    parchment_percent,
    moisture_loss_percent,
    parchment_allowed,
  })
  const has_recipe = nominal_tube_paper_g > 0 && expanded.length > 0
  // A selected paper recipe is physical truth. Target weight is a comparison
  // benchmark and must never scale GSM/geometry-derived paper consumption.
  const paper_calibration_factor = has_recipe ? 1 : 0
  const target_per_ply_wpm = [...per_ply_wpm]
  const target_paper_wpm = target_per_ply_wpm.reduce((sum, weight) => sum + weight, 0)
  const tube = has_recipe
    ? wetDryBreakdown(nominal_tube_paper_g, {
        adhesive_percent,
        parchment_percent,
        moisture_loss_percent,
        parchment_allowed,
        target_dry_g: opts.target_dry_g,
      })
    : { paper_g: 0, adhesive_g: 0, parchment_g: 0, wet_g: 0, dry_g: 0 }

  const plan = buildBambooPlan(tube_length_mm || 1)
  const tube_basis_mm = Math.max(tube_length_mm, 1)
  const bamboo = scaleBreakdown(tube, plan.tubes_per_bamboo)
  const bamboo_trim = scaleBreakdown(tube, plan.total_trim_mm / tube_basis_mm)
  const whole_bamboo = scaleBreakdown(tube, plan.bamboo_length_mm / tube_basis_mm)

  const validation = validateRecipe(opts.papers, opts.target_dry_g || 0, tube.dry_g)

  return {
    id_mm: round4(id_mm),
    mandrel_tolerance_mm: MANDREL_TOLERANCE_MM,
    od_mm: round4(od_mm),
    wall_mm: round4(wall_mm),
    per_ply_thickness_mm: thicknesses.map(round4),
    per_ply_avg_dia_mm: avg_dias.map(round4),
    per_ply_weight_per_mm_g: per_ply_wpm.map(round6),
    target_per_ply_weight_per_mm_g: target_per_ply_wpm.map(round6),
    paper_weight_per_mm_g: round6(paper_wpm),
    target_paper_weight_per_mm_g: round6(target_paper_wpm),
    tube,
    nominal_tube,
    bamboo,
    bamboo_trim,
    whole_bamboo,
    bamboo_plan: plan,
    paper_required_g: required,
    paper_calibration_factor: round6(paper_calibration_factor),
    validation,
  }
}
